src/helpers/ControlRecordTreeBuilder.js

```diff
@@ -55,10 +59,45 @@ export const ControlRecord = new Record({
     }
     return value;
   },
+  remove(formFieldPath) {
+    function findIfNodeHasDataSourceWithValueInDatabase(parent) {
+      if (parent.dataSource.uuid
+        || (parent.dataSource.obsList &&
+          parent.dataSource.obsList.find(o => o.uuid !== undefined))) return true;
+
+      if (parent.children) {
+        for (let i = 0; i < parent.children.size; i++) {
+          if (findIfNodeHasDataSourceWithValueInDatabase(parent.children.get(i))) return true;
+        }
+      }
+      return false;
+    }
+    if (this.children) {
+      let updatedChildren = this.children.filter(child => child.formFieldPath !== formFieldPath);
+      if (updatedChildren.size === this.children.size) {
+        updatedChildren = this.children.map((child) => child.remove(formFieldPath));
+      } else {
+        const removedChild = this.children.find(child => child.formFieldPath === formFieldPath);
+        if (findIfNodeHasDataSourceWithValueInDatabase(removedChild) && removedChild) {
+          let voidedChildRecord = removedChild.set('active', false).set('voided', true)
+            .voidChildRecords();
+          if (voidedChildRecord.dataSource.obsList) {
+            const newDataSource = voidedChildRecord.dataSource.set('obsList',
+              voidedChildRecord.dataSource.obsList.map(o => o.set('voided', true)));
+            voidedChildRecord = voidedChildRecord.set('dataSource', newDataSource);
+          }
+          updatedChildren = updatedChildren.insert(updatedChildren.size, voidedChildRecord);
+        }
+      }
+      return this.set('children', updatedChildren);
+    }
+    return this;
+  },
 
   update(formFieldPath, value, errors, isRemoved) {
     if (this.formFieldPath === formFieldPath) {
-      return (Object.keys(value).length === 0 && isRemoved ? this.set('active', false) : this)
+      return (Object.keys(value).length === 0 && isRemoved ? this.set('active', false)
```

Comment:
@shivarach @bsneha90 can we remove **isRemoved** parameter as its not being used with the latest changes

Reply:
removed and committed